File: ruby/src/general/longest_palindromic_substring.rb
module General
  class << self

    # Given a string S, find the longest palindromic substring in S.
    def get_longest_palindromic_substring(str)
      return str if !str || str.size == 0

      max_start_index = max_end_index = -1
      (0..str.size - 1).each do |index|
        max_start_index, max_end_index = do_longest_palindromic_substring(str,
                                                                          index - 1, index + 1,
                                                                          max_start_index, max_end_index)
        max_start_index, max_end_index = do_longest_palindromic_substring(str,
                                                                          index, index + 1,
                                                                          max_start_index, max_end_index)
      end

      str[max_start_index..max_end_index]
    end

    def do_longest_palindromic_substring(str, start_index, end_index,
                                         max_start_index, max_end_index)
      while start_index >= 0 && end_index <= str.size - 1 &&
          str[start_index] == str[end_index]
        start_index -= 1
        end_index += 1
      end

      start_index += 1
      end_index -= 1
      return [start_index, end_index] if exceeded?(start_index, end_index,
                                                   max_start_index, max_end_index)
      [max_start_index, max_end_index]
    end

    def exceeded?(start_index, end_index, max_start_index, max_end_index)
      (max_end_index == -1 || max_start_index == -1) ||
        ((end_index - start_index + 1) > (max_end_index - max_start_index + 1))
    end
  end
end

File: java/src/general/LongestPalindromicSubstring.java
package general;

import common.Pair;

/**
 * Given a string S, find the longest palindromic substring in S.
 */
public class LongestPalindromicSubstring {
	public static String getLongestPalindromicSubstring(String s) {
		if (s == null || s.isEmpty()) {
			return s;
		}

		Pair<Integer, Integer> maxPalindrome = new Pair<Integer, Integer>();
		Pair<Integer, Integer> current = new Pair<Integer, Integer>();
		for (int index = 0; index < s.length(); index++) {
			current.setFirst(index - 1);
			current.setSecond(index + 1);
			getLongestPalindromicSubstring(s, current, maxPalindrome);

			current.setFirst(index);
			current.setSecond(index + 1);
			getLongestPalindromicSubstring(s, current, maxPalindrome);
		}

		return s.substring(maxPalindrome.getFirst(),
				maxPalindrome.getSecond() + 1);
	}

	private static void getLongestPalindromicSubstring(String s,
			Pair<Integer, Integer> current, Pair<Integer, Integer> maxPalindrome) {
		int startIndex = current.getFirst(), endIndex = current.getSecond();
		while (startIndex >= 0 && endIndex <= s.length()
				&& s.charAt(startIndex) == s.charAt(endIndex)) {
			startIndex--;
			endIndex++;
		}

		current.setFirst(startIndex);
		current.setSecond(endIndex);
		if (exceeded(current, maxPalindrome)) {
			maxPalindrome.setFirst(current.getFirst());
			maxPalindrome.setSecond(current.getSecond());
		}
	}

	private static boolean exceeded(Pair<Integer, Integer> current,
			Pair<Integer, Integer> maxPalindrome) {
		int startIndex = current.getFirst(), endIndex = current.getSecond();
		int maxStartIndex = maxPalindrome.getFirst(), maxEndIndex = maxPalindrome
				.getSecond();

		if (maxStartIndex == -1
				|| (endIndex - startIndex + 1) > (maxEndIndex - maxStartIndex + 1)) {
			return true;
		}

		return false;
	}
}
